pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+        pub fn update_validators_pool(validator: T::AccountId, total: BalanceOf<T>) {
+            log::trace!(
+                "update_validators_pool:[{:#?}] | Own[{:#?}] | Tot[{:#?}]",
+                line!(),
+                validator,
+                total,
+            );
+            <ValidatorPool<T>>::mutate(|validators| {
+                validators.remove(&Bond::from_owner(validator.clone()));
+                validators.insert(Bond {
+                    owner: validator,
+                    amount: total,
+                });
+            });
+        }
+        // ensure validator is active before calling
+        pub fn remove_from_validators_pool(validator: T::AccountId) {
+            log::trace!(
+                "remove_from_validators_pool:[{:#?}] | Own[{:#?}]",
+                line!(),
+                validator
+            );
+            <ValidatorPool<T>>::mutate(|validators| {
+                validators.remove(&Bond::from_owner(validator.clone()));
+            });
+        }
+        pub(crate) fn validator_deactivate(controller: &T::AccountId) {
+            log::trace!(
+                "validator_deactivate:[{:#?}] - Acc[{:#?}]",
+                line!(),
+                controller
+            );
+            <ValidatorState<T>>::mutate(&controller, |maybe_validator| {
+                if let Some(valid_state) = maybe_validator {
+                    valid_state.go_offline();
+                    Self::remove_from_validators_pool(controller.clone());
+                }
+            });
+        }
+        fn nominator_leaves_validator(
+            nominator: T::AccountId,
+            validator: T::AccountId,
+        ) -> DispatchResultWithPostInfo {
+            <ValidatorState<T>>::try_mutate_exists(
+                validator.clone(),
+                |maybe_validator| -> DispatchResultWithPostInfo {
+                    let mut state = maybe_validator.as_mut().ok_or(<Error<T>>::ValidatorDNE)?;
+                    let mut exists: Option<BalanceOf<T>> = None;
+                    let noms = state
+                        .clone()
+                        .nominators
+                        .0
+                        .into_iter()
+                        .filter_map(|nom| {
+                            if nom.owner != nominator {
+                                Some(nom)
+                            } else {
+                                exists = Some(nom.amount);
+                                None
+                            }
+                        })
+                        .collect();
+                    let nominator_stake = exists.ok_or(<Error<T>>::ValidatorDNE)?;
+                    let nominators = OrderedSet::from(noms);
+
+                    // T::Currency::unreserve(&nominator, nominator_stake);
+
+                    state.nominators = nominators;
+                    state.nomi_bond_total -= nominator_stake;
+                    state.total -= nominator_stake;
```

Comment:
use safe math or saturating operators :)

Reply:
:+1:  Moved to saturating_sub()